GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m

```diff
@@ -330,30 +352,100 @@ - (instancetype)init {
+  NSMutableDictionary<NSNumber *, dispatch_semaphore_t> *map = sBackgroundIdentifierToSemaphoreMap;
+  if (queue && map) {
+    dispatch_sync(queue, ^{
+      bgID = arc4random();
+      while (bgID == GDTCORBackgroundIdentifierInvalid || map[@(bgID)]) {
+        bgID = arc4random();
+      }
+      map[@(bgID)] = semaphore;
+    });
+  }
+  return bgID;
+}
+
+/** Returns the semaphore mapped to given bgID and removes the value from the map.
+ *
+ * @param bgID The unique NSUInteger as GDTCORBackgroundIdentifier.
+ * @return The semaphore mapped by given bgID.
+ */
++ (dispatch_semaphore_t)semaphoreForBackgroundIdentifier:(GDTCORBackgroundIdentifier)bgID {
+  __block dispatch_semaphore_t semaphore;
+  dispatch_queue_t queue = sSemaphoreQueue;
+  NSMutableDictionary<NSNumber *, dispatch_semaphore_t> *map = sBackgroundIdentifierToSemaphoreMap;
+  if (queue && map) {
+    dispatch_sync(queue, ^{
+      semaphore = map[@(bgID)];
+      [map removeObjectForKey:@(bgID)];
+    });
+  }
+  return semaphore;
+}
+
 - (GDTCORBackgroundIdentifier)beginBackgroundTaskWithName:(NSString *)name
                                         expirationHandler:(void (^)(void))handler {
+  __block GDTCORBackgroundIdentifier bgID = GDTCORBackgroundIdentifierInvalid;
 #if !TARGET_OS_WATCH
-  GDTCORBackgroundIdentifier bgID =
-      [[self sharedApplicationForBackgroundTask] beginBackgroundTaskWithName:name
-                                                           expirationHandler:handler];
+  bgID = [[self sharedApplicationForBackgroundTask] beginBackgroundTaskWithName:name
+                                                              expirationHandler:handler];
 #if !NDEBUG
   if (bgID != GDTCORBackgroundIdentifierInvalid) {
-    GDTCORLogDebug("Creating background task with name:%@ bgID:%ld", name, (long)bgID);
+    GDTCORLogDebug("Creating background task with name:%@ bgID:%ld.", name, (long)bgID);
   }
 #endif  // !NDEBUG
+#elif TARGET_OS_WATCH
+  dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+  bgID = [GDTCORApplication createAndMapBackgroundIdentifierToSemaphore:semaphore];
+  if (bgID != GDTCORBackgroundIdentifierInvalid) {
+    GDTCORLogDebug(@"Creating activity with name:%@ bgID:%ld on watchOS.", name, (long)bgID);
+  }
+  [[self sharedNSProcessInfoForBackgroundTask]
+      performExpiringActivityWithReason:name
+                             usingBlock:^(BOOL expired) {
+                               if (expired) {
+                                 if (handler) {
+                                   handler();
+                                 }
+                                 dispatch_semaphore_signal(semaphore);
+                                 GDTCORLogDebug(
+                                     @"Activity with name:%@ bgID:%ld on watchOS is expiring.",
+                                     name, (long)bgID);
+                               }
+                               dispatch_semaphore_wait(
```

Comment:
This needs to be put into an else block, or the if (expired) block above needs to have a return call in there. Otherwise, dispatch_semaphore_wait will always be called and it'll freeze the app or force it to be terminated.

Reply:
Fixed